kyuubi-server/src/main/scala/org/apache/kyuubi/client/KyuubiSyncThriftClient.scala

```diff
@@ -69,27 +140,49 @@ class KyuubiSyncThriftClient private (protocol: TProtocol)
     req.setUsername(user)
     req.setPassword(password)
     req.setConfiguration(configs.asJava)
-    val resp = withLockAcquired(OpenSession(req))
+    val resp = withRetryingRequest(OpenSession(req), "OpenSession")
     ThriftUtils.verifyTStatus(resp.getStatus)
     _remoteSessionHandle = resp.getSessionHandle
     _engineId = Option(resp.getConfiguration)
       .filter(_.containsKey("kyuubi.engine.id"))
       .map(_.get("kyuubi.engine.id"))
+
+    engineAliveProbeClient.foreach { aliveProbeClient =>
+      Utils.tryLogNonFatalError {
+        val resp = aliveProbeClient.OpenSession(req)
+        ThriftUtils.verifyTStatus(resp.getStatus)
+        _aliveProbeSessionHandle = resp.getSessionHandle
+        startEngineAliveProbe()
+      }
+    }
+
     SessionHandle(_remoteSessionHandle, protocol)
   }
 
   def closeSession(): Unit = {
     try {
       if (_remoteSessionHandle != null) {
         val req = new TCloseSessionReq(_remoteSessionHandle)
-        val resp = withLockAcquired(CloseSession(req))
+        val resp = withRetryingRequest(CloseSession(req), "CloseSession")
         ThriftUtils.verifyTStatus(resp.getStatus)
       }
     } catch {
       case e: Exception =>
         throw KyuubiSQLException("Error while cleaning up the engine resources", e)
     } finally {
-      if (protocol.getTransport.isOpen) protocol.getTransport.close()
+      Seq(protocol).union(engineAliveProbeProtocol.toSeq).foreach { tProtocol =>
+        if (tProtocol.getTransport.isOpen) tProtocol.getTransport.close()
+      }
+      Option(engineAliveThreadPool).foreach(_.shutdown())
+      if (_aliveProbeSessionHandle != null && !remoteEngineBroken) {
+        engineAliveProbeClient.foreach { client =>
+          Utils.tryLogNonFatalError {
+            val req = new TCloseSessionReq(_aliveProbeSessionHandle)
+            val resp = client.CloseSession(req)
```

Comment:
the odering seems wrong, you has closed the transport.

Reply:
thanks